cmake/modules/RootNewMacros.cmake

```diff
@@ -282,7 +282,34 @@ function(ROOT_GENERATE_DICTIONARY dictionary)
       if(IS_ABSOLUTE ${fp})
         set(headerFile ${fp})
       else()
-        find_file(headerFile ${fp} HINTS ${incdirs} NO_DEFAULT_PATH NO_SYSTEM_ENVIRONMENT_PATH)
+        set(incdirs_in_build)
+        set(incdirs_in_prefix)
+        foreach(incdir ${incdirs})
+          if(NOT IS_ABSOLUTE incdir
+             OR incdir MATCHES "^${CMAKE_SOURCE_DIR}"
+             OR incdir MATCHES "^${CMAKE_CURRENT_BUILD_DIR}"
+             OR incdir MATCHES "^${CMAKE_BUILD_DIR}")
+            list(APPEND incdirs_in_build
+                 ${incdir})
+          else()
+            list(APPEND incdirs_in_prefix
+                 ${incdir})
+          endif()
+        endforeach()
+        if(incdirs_in_build)
+          find_file(headerFile ${fp}
+            HINTS ${incdirs_in_build}
+            NO_DEFAULT_PATH
+            NO_SYSTEM_ENVIRONMENT_PATH
+            NO_CMAKE_FIND_ROOT_PATH)
+        endif()
+        # Try this even if NOT incdirs_in_prefix: might not need a HINT.
+        if(NOT headerFile)
+          find_file(headerFile ${fp}
+            HINTS ${incdirs_in_prefix}
+            NO_DEFAULT_PATH
+            NO_SYSTEM_ENVIRONMENT_PATH)
+        endif()
```

Comment:
Should we encapsulate this into its own function, maybe?

Reply:
Can do - do you have a suggestion for a name? I can only come up with `Find_file_with_CMAKE_FIND_ROOT_PATH_or_not` ;-) I'm also not entirely sure whether that improves the function a lot - CMake isn't exactly C++...

Reply:
@amadio decided this can go in as is.